# Objetivo del proyecto 


Analizar un conjunto de datos de accidentes de tráfico en Barcelona y construir un modelo que prediga la gravedad del accidente (Leve, Grave o Mortal) usando variables como tipo de vehículo, tipo de accidente, edad, sexo, distrito, condiciones meteorológicas, etc.

## Fase 1

Durante la fase 1 del proyecto procederé a la importación de datos de Accidentes gestionados por la guardia urbana de Barcelona de 2024 (Fuente: https://opendata-ajuntament.barcelona.cat/data/es/dataset/accidents-gu-bcn) usando Pandas. Centraré esta fase en la Limpieza, EDA (Exploratory Data Analysis), visualización (con Matplotlb y Seaborn) y transformación de los datos

### 1. Carga y exploración inicial de datos

In [1]:
# Importamos las librerías que vamos a utilizar durante esta fase
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
# Establecemos la configuración para visualizar los DF de pandas 
pd.options.display.min_rows=6
%config IPCompleter.greedy=True

Una vez importadas las librerías procedemos a cargar el DataSet correspondiente al año 2024 y lo guardamos en la variable **accidentes**. Creo la variable ruta donde meter la ruta donde se guarde el archivo y porder abrirlo desde ahí.

In [3]:
ruta='2024_accidents_gu_bcn.csv'
accidentes=pd.read_csv(ruta)

#Saco la info del DataFrame para ver nombre de variables, tipo de cada variable, ver número total de registros, numero de variables y conteo de no nulos
accidentes.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7536 entries, 0 to 7535
Data columns (total 25 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Numero_expedient           7536 non-null   object 
 1   Codi_districte             7536 non-null   int64  
 2   Nom_districte              7536 non-null   object 
 3   Codi_barri                 7536 non-null   int64  
 4   Nom_barri                  7536 non-null   object 
 5   Codi_carrer                7536 non-null   int64  
 6   Nom_carrer                 7536 non-null   object 
 7   Num_postal                 4425 non-null   object 
 8   Descripcio_dia_setmana     7536 non-null   object 
 9   NK_Any                     7536 non-null   int64  
 10  Mes_any                    7536 non-null   int64  
 11  Nom_mes                    7536 non-null   object 
 12  Dia_mes                    7536 non-null   int64  
 13  Hora_dia                   7536 non-null   int64

En este DataSet tiene sentido que el indice sea el número de expediente. 

In [4]:
accidentes.set_index('Numero_expedient', inplace=True)
accidentes

,Codi_districte,Nom_districte,Codi_barri,Nom_barri,Codi_carrer,Nom_carrer,Num_postal,Descripcio_dia_setmana,NK_Any,Mes_any,...,Descripcio_causa_vianant,Numero_morts,Numero_lesionats_lleus,Numero_lesionats_greus,Numero_victimes,Numero_vehicles_implicats,Coordenada_UTM_Y_ED50,Coordenada_UTM_X_ED50,Longitud_WGS84,Latitud_WGS84
Numero_expedient,,,,,,,,,,,,,,,,,,,,,
2024S000001,9,Sant Andreu,57,la Trinitat Vella,701268,Dalt (Llobregat),K165,Dilluns,2024,1,...,NaN,NaN,2.0,NaN,2.0,1,432839.501,4589349.921,2.194806,41.451076
2024S000002,2,Eixample,8,l'Antiga Esquerra de l'Eixample,144601,Diagonal,423-425,Dilluns,2024,1,...,NaN,NaN,1.0,NaN,1.0,2,429284.086,4583117.055,2.152973,41.394633
2024S000003,2,Eixample,10,Sant Antoni,169409,Corts Catalanes,NaN,Dilluns,2024,1,...,NaN,NaN,4.0,1.0,5.0,2,429834.070,4581818.313,2.159702,41.382984
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2024S007549,3,Sants-Montjuïc,11,el Poble Sec,198804,Paral·lel,NaN,Dijous,2024,9,...,NaN,NaN,1.0,NaN,1.0,1,428989.313,4580940.290,2.149703,41.375002
2024S007550,3,Sants-Montjuïc,11,el Poble Sec,234001,Olímpic,1-3,Dilluns,2024,11,...,NaN,NaN,1.0,NaN,1.0,1,429206.725,4579426.054,2.152479,41.361383
2024S007551,1,Ciutat Vella,1,el Raval,701367,Raval,25X,Divendres,2024,12,...,NaN,NaN,1.0,NaN,1.0,1,430668.432,4581341.695,2.169734,41.378764


In [14]:
accidentes.describe(include= ['object']).T

,count,unique,top,freq
Nom_districte,7532,10,Eixample,1898
Nom_barri,7532,73,la Dreta de l'Eixample,688
Nom_carrer,7532,979,Corts Catalanes,385
Num_postal,4421,1397,1,92
Descripcio_dia_setmana,7532,7,Dimarts,1223
Nom_mes,7532,12,Juliol,712
Descripcio_torn,7532,3,Tarda,3589
Descripcio_causa_vianant,471,5,Desobeir el senyal del semàfor,172


In [15]:
accidentes.describe(exclude= ['object']).T

,count,mean,std,min,25%,50%,75%,max
Codi_districte,7532.0,5.037971e+00,2.981623,1.000000e+00,2.000000e+00,5.000000e+00,8.000000e+00,1.000000e+01
Codi_barri,7532.0,2.837281e+01,22.044700,1.000000e+00,8.000000e+00,2.300000e+01,4.400000e+01,7.300000e+01
Codi_carrer,7532.0,2.614639e+05,208680.156534,1.800000e+02,1.228302e+05,1.973020e+05,3.365082e+05,7.019180e+05
NK_Any,7532.0,2.024000e+03,0.000000,2.024000e+03,2.024000e+03,2.024000e+03,2.024000e+03,2.024000e+03
Mes_any,7532.0,6.547929e+00,3.443103,1.000000e+00,4.000000e+00,7.000000e+00,1.000000e+01,1.200000e+01
Dia_mes,7532.0,1.566423e+01,8.733805,1.000000e+00,8.000000e+00,1.600000e+01,2.300000e+01,3.100000e+01
Hora_dia,7532.0,1.366052e+01,5.390857,0.000000e+00,1.000000e+01,1.400000e+01,1.800000e+01,2.300000e+01
Numero_morts,11.0,1.000000e+00,0.000000,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00
Numero_lesionats_lleus,6525.0,1.263295e+00,0.685720,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.500000e+01
Numero_lesionats_greus,241.0,1.053942e+00,0.366849,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,6.000000e+00


### 2. Limpieza de Datos (Data Cleaning)

#### 2.1 Gestión de duplicados

Cuento todos los duplicados y los muestro por pantalla, para chequear potenciales errores.

In [7]:
accidentes.duplicated(keep=False).sum()

np.int64(8)

Hay un total de 4 registros que parecen duplicados. Tienen distinto numero de expediente pero los datos del accidente son iguales (coinciden en dia y fecha con el único cambio del nº de expediente). Voy a asumir que son debidos a duplicidades en la toma de datos, con lo que me quedaré solo con el primer valor que aparece. Hago la comprobación de que no hay duplicados tras eliminarlos y listo el DF para ver nº de registros (4 menos que el aterior 7536 vs 7532)

In [7]:
accidentes[accidentes.duplicated(keep=False)]

,Codi_districte,Nom_districte,Codi_barri,Nom_barri,Codi_carrer,Nom_carrer,Num_postal,Descripcio_dia_setmana,NK_Any,Mes_any,...,Descripcio_causa_vianant,Numero_morts,Numero_lesionats_lleus,Numero_lesionats_greus,Numero_victimes,Numero_vehicles_implicats,Coordenada_UTM_Y_ED50,Coordenada_UTM_X_ED50,Longitud_WGS84,Latitud_WGS84
Numero_expedient,,,,,,,,,,,,,,,,,,,,,
2024S000467,1,Ciutat Vella,3,la Barceloneta,197903,Marítim de la Barceloneta,23-31,Divendres,2024,1,...,NaN,NaN,1.0,NaN,1.0,2,432758.379,4581836.017,2.194671,41.383394
2024S000550,1,Ciutat Vella,3,la Barceloneta,197903,Marítim de la Barceloneta,23-31,Divendres,2024,1,...,NaN,NaN,1.0,NaN,1.0,2,432758.379,4581836.017,2.194671,41.383394
2024S002432,6,Gràcia,29,el Coll,216702,Mons,6,Dilluns,2024,4,...,NaN,NaN,NaN,NaN,NaN,1,428631.844,4585219.869,2.144924,41.413515
2024S002436,6,Gràcia,29,el Coll,216702,Mons,6,Dilluns,2024,4,...,NaN,NaN,NaN,NaN,NaN,1,428631.844,4585219.869,2.144924,41.413515
2024S003470,5,Sarrià-Sant Gervasi,27,el Putxet i el Farró,353408,Vallirana,30,Diumenge,2024,6,...,NaN,NaN,1.0,NaN,1.0,1,428918.525,4584119.180,2.148483,41.403627
2024S003585,5,Sarrià-Sant Gervasi,27,el Putxet i el Farró,353408,Vallirana,30,Diumenge,2024,6,...,NaN,NaN,1.0,NaN,1.0,1,428918.525,4584119.180,2.148483,41.403627
2024S006447,3,Sants-Montjuïc,11,el Poble Sec,234001,Olímpic,1-3,Dilluns,2024,11,...,NaN,NaN,1.0,NaN,1.0,1,429206.725,4579426.054,2.152479,41.361383
2024S007550,3,Sants-Montjuïc,11,el Poble Sec,234001,Olímpic,1-3,Dilluns,2024,11,...,NaN,NaN,1.0,NaN,1.0,1,429206.725,4579426.054,2.152479,41.361383


In [9]:
accidentes.drop_duplicates(keep='first', inplace=True)
accidentes.duplicated().sum()

np.int64(0)

#### 2.2 Conversión de tipos de datos

Tras analizar los tipos de datos, tiene sentido incluir desde código de distrito hasta descripción de causa como variables categóricas y dejar el resto como variables numéricas. Procedo a cambiar el tipo de variable según lo comentado, y lo compruebo por pantalla

In [21]:
col=accidentes.loc[:,'Codi_districte':'Descripcio_causa_vianant'].columns
accidentes[col]=accidentes[col].astype('category')
accidentes.dtypes


Codi_districte               category
Nom_districte                category
Codi_barri                   category
Nom_barri                    category
Codi_carrer                  category
Nom_carrer                   category
Num_postal                   category
Descripcio_dia_setmana       category
NK_Any                       category
Mes_any                      category
Nom_mes                      category
Dia_mes                      category
Hora_dia                     category
Descripcio_torn              category
Descripcio_causa_vianant     category
Numero_morts                  float64
Numero_lesionats_lleus        float64
Numero_lesionats_greus        float64
Numero_victimes               float64
Numero_vehicles_implicats       int64
Coordenada_UTM_Y_ED50         float64
Coordenada_UTM_X_ED50         float64
Longitud_WGS84                float64
Latitud_WGS84                 float64
dtype: object

In [22]:
accidentes.describe(include='category').T

,count,unique,top,freq
Codi_districte,7532,10,2,1898
Nom_districte,7532,10,Eixample,1898
Codi_barri,7532,73,7,688
Nom_barri,7532,73,la Dreta de l'Eixample,688
Codi_carrer,7532,1015,169409,385
Nom_carrer,7532,979,Corts Catalanes,385
Num_postal,4421,1397,1,92
Descripcio_dia_setmana,7532,7,Dimarts,1223
NK_Any,7532,1,2024,7532
Mes_any,7532,12,7,712


#### 2.3 Gestión de valores nulos o faltantes

A continuación saco un listado de valores nulos y lo ordeno de manera descendente

In [23]:
accidentes.isna().sum().sort_values(ascending=False)

Numero_morts                 7521
Numero_lesionats_greus       7291
Descripcio_causa_vianant     7061
Num_postal                   3111
Numero_lesionats_lleus       1007
Numero_victimes               815
Codi_barri                      0
Codi_districte                  0
Descripcio_dia_setmana          0
Nom_carrer                      0
Codi_carrer                     0
Nom_barri                       0
Nom_districte                   0
Dia_mes                         0
NK_Any                          0
Mes_any                         0
Hora_dia                        0
Descripcio_torn                 0
Nom_mes                         0
Numero_vehicles_implicats       0
Coordenada_UTM_Y_ED50           0
Coordenada_UTM_X_ED50           0
Longitud_WGS84                  0
Latitud_WGS84                   0
dtype: int64

Para la eliminación de nulos asumiremos los siguientes supuestos:
1. **Número de muertos, Número de lesionados graves y número de lesionados leves**: los nulos los sustituiremos por el valor 0. Tiene sentido que la guardia urbana haya metido N/A en el caso de que no haya muertos ni lesionados de ninguna consideración
2. **Num postal, descripción de causa vianant**: crearemos una nueva categoría 'No disponible' y cambiaremos los nulos por valor 'No disponible'
3. **Numero_victimes**: Tiene sentido remplazar el valores nulos por el valor que mas se repite (moda). No reemplazo estos valores por 0 porque siempre en un accidente hay victimas (al menos 1).

#### 2.4 Detección y corrección de errores

### 3. Análisis de Calidad de datos

### 4 EDA (Analisis Exploratorio de Datos)